modules/phase_field/src/postprocessors/FeatureFloodCount.C

```diff
@@ -467,129 +468,76 @@ FeatureFloodCount::mergeSets(bool use_periodic_boundary_info)
-              it2->_merged = true;
-
-              // Something was merged so we'll need to repeat this loop
-              region_merged = true;
-            }
-          } // it2 loop
-
-          // Don't increment if we had a merge, we need to retry earlier candidates again
-          if (!region_merged)
-            ++it1;
-          else
-            goto REPEAT_MERGE_LOOPS;
-
-        } // it1 loop
-      } // rank2 loop
-    } // rank1 loop
-  } // map loop
+             )
+           )
+        {
+          it2->merge(std::move(*it1));
 
+          // Insert the new entity at the end of the list so that it may be checked against all other partial features again
+          _partial_feature_sets[map_num].emplace_back(std::move(*it2));
 
-  for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
-    _feature_sets[map_num].clear();
+          /**
+           * Now remove both halves the merged features: it2 contains the "moved" feature cell just inserted
+           * at the back of the list, it1 contains the mostly empty other half. We have to be careful about the
+           * order in which these two elements are deleted. We delete it2 first since we don't care where its
+           * iterator points after the deletion. We are going to break out of this loop anyway. If we delete
+           * it1 first, it may end up pointing at the same location as it2 which after the second deletion would
+           * cause both of the iterators to be invalidated.
+           */
+          _partial_feature_sets[map_num].erase(it2);
+          it1 = _partial_feature_sets[map_num].erase(it1); // it1 is incremented here!
 
-  // Now consolidate the data structure
-  _feature_count = 0;
-  for (auto rank = decltype(_n_procs)(0); rank < _n_procs; ++rank)
-    for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
-    {
-      for (auto & feature : _partial_feature_sets[rank][map_num])
-      {
-        if (!feature._merged)
-        {
-          _feature_sets[map_num].emplace_back(std::move(feature));
-          ++_feature_count;
+          // A merge occurred, this is used to determine whether or not we increment the outer iterator
+          merge_occured = true;
+
+          // We need to start the list comparison over for the new it1 so break here
+          break;
         }
-      }
+      } // it2 loop
+
+      if (!merge_occured) // No merges so we need to manually increment the outer iterator
+        ++it1;
+
+    } // it1 loop
+  } // map loop
 
-      _partial_feature_sets[rank][map_num].clear();
+  /**
+   * All of the merges are complete and stored in a vector of lists. To make several
+   * of the sorting and tracking algorithms more straightforward, we will move these
+   * items into a vector of vectors instead.
+   */
+  _feature_count = 0;
+  for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
```

Comment:
And the problem is... if you screw up and assign a `bar_id_type` to a `foo_id_type` they might be the same right now and not give any compiler warnings and work just fine so you would never know... and then later you change one of them and you have a bug.

_This_ line of code is foolproof.